### ISSUE TITLE
UrlMatcher overly lenient with empty params

### ISSUE THREAD

**Issue by kevinkyyro** [criteo/lolhttp#55]
I found this unintuitive, though it may or may not actually be correct (haven't read the corresponding RFCs) and I apologize in advance if it is.

 A `UrlMatcher` pattern containing a query parameter that has no value (`/a?b=`) matches urls that do not have the parameter present at all.

So, the following works, though it seems to me that it shouldn't.

```scala
import io.http._

val url"/?section=&update=${date}" = "/?section=contact&update=now"
// date: String = now
```

**Comment by Masuzu** [comment]
According to https://tools.ietf.org/html/rfc3986#section-3, the hierarchical path preceding the query parameters can be an empty path, isn't it?

**Comment by kevinkyyro** [comment]
The question is about the query, not the path. Seems like it shouldn't match without the section param being empty as well. Like `url"/?section=${""}"`

**Comment by guillaumebort** [comment]
Hum yes I agree, if the value parameter is not a variable value specified as `${}` it should only match exact match.

**Comment by guillaumebort** [comment]
Parameters in queryString are optional if they are defined

**Comment by guillaumebort** [comment]
Ok it should be fixed by PR #59 

A parameter matching a dynamic expression like `section=${section}` will match everything, including if the parameter is missing (in this case the matched value will be ""{empty string}). It is important otherwise you can't really create handler with optional parameters.

A parameter value set in a literal value in the matcher expression (including ""{empty string}) will always be an exact match.
